### ISSUE TITLE
Query state changes based on block index

### ISSUE THREAD

**Issue by ilblackdragon** [nearprotocol/nearcore#1600]
State changes (which replace events) should be queryable by block index / block hash for the past blocks.

**Comment by ailisp** [comment]
is state change already stored in block right now?

**Comment by bowenwang1996** [comment]
No. I think this is for events.

**Comment by ailisp** [comment]
related to #1546. 3. I think it's coupled on how state update is storing and when is it send to subscriber. @frol would you plan to 1) send event to subscriber right at the moment a state change is commited, and then runtime store it. or 2) store it first in a kv db, and you listen to db changes?
If plan is close to first, I can do store while you do subscription in parallel.

@bowenwang1996 since it's not in block yet, is it planned to stored in block, or a out of chain storage.

**Comment by ilblackdragon** [comment]
@frol 
So we need this first without any subscriptions.
E.g. we should be able to query a state change for given record in the storage for given block:
```
http post URL method:=changes params:='["/path/to/state", block_hash]' ...
```

Note, that we need to also be able to query NEAR balance changes like this as well.

**Comment by frol** [comment]
@mikhailOK I have heard that the storage was under a major rewrite. Is it complete? Could you give any helpers where I can start with getting the information to expose via RPC? (I know the RPC side of things, but I am still wrapping my head around the internals)

@ilblackdragon Do you have an exampe on how `"/path/to/state"` might look like?

**Comment by nearmax** [comment]
So this is different from the model where we first subscribe to certain keys and later get the recorded changes. Just FYI, the subscription model consists of the following components:
- [ ] Exposing changes in `TrieUpdate` and in `ApplyResult` https://github.com/nearprotocol/nearcore/pull/1652 (it was implemented and not rolled back and it is not affected by the storage rewrite);
- [ ] Events actor that collects events from the client actor from the `ApplyResult` and sends it over WAMP;
- [ ] WAMP.

The feature in this issue is different. It should allow retrieving the state changes for the given prefixes and the given block range. This feature is difficult, since we don't want to replay the blocks and we currently do not store old versions of the state in the storage. We should change the storage such that we store several past versions of the state and use it to reply to RPC, unless RPC requests events very far in the past in which case we reply the blocks. Unfortunately, since we don't want to replay the blocks from the genesis we need to periodically save the state for blocks far in the past.
MVP:
- [ ] RPC interface to request the changes;
- [ ] Change the storage to keep several past states. RPC then would read from this history and return an error if the RPC request goes too far in the past.
Full implementation:
- [ ] Checkpoint state every several blocks, such that the further the blocks are in the past the less frequent they are checkpointed;
- [ ] When RPC requests state changes from very far in the past find the checkpoint and replay the blocks starting from it.

**Comment by ilblackdragon** [comment]
Ok, let's make sure we implement the RPC API to be:

```
http post URL method:=changes params:='["/path/to/state", block_hash]'
```

where /path/to/state is for example:
 - / - for all changes
 -  /illia.china for account `illia.china`
 - /contract/key_in_contracts_data for `key_in_contracts_data` for `contract` storage. keys can have more / or other separators.

also let's support `params:='["/path/to/state", "latest"]'` for latest (finalized?) block.

**Comment by nearmax** [comment]
@ilblackdragon Could you also provide a format of the response?

**Comment by nearmax** [comment]
@frol Here are the discussed work items:

* Replace `TrieUpdate::committed` with a `Vec<(Vec<u8>, CryptoHash, Option<Vec<u8>>)>` or `BTreeMap<Vec<u8>, Vec<(CryptoHash, Option<Vec<u8>>)>>`, where `CryptoHash` is the has of the receipt that introduced the change;
* Change `TrieUpdate::commit` to accept the `CryptoHash` of the receipt/transaction that introduced the change;
* In `TrieUpdate::commit` instead of overwriting values in `self.commit` from `self.prospective`, we need to extend it;
* In `TrieUpdate::finalize` iterate over `self.committed` and dump the values into a new column in the database, e.g. `COL_KEY_CHANGES`, you could store it as concat(block_index, key) -> (transaction, new_value);
* In RPC, we need to iterate over this column and retrieve the pairs `(transaction, new_value)` given the range of block indecies and the key. This can be achieved e.g. by extending `ViewRuntimeAdapter` trait and accessing `store` from `NightshadeRuntime`.

**Comment by frol** [comment]
The updated design is to keep `TrieUpdate` implementation storing `Vec<u8>` values by `Vec<u8>` keys the way it used to be, but extend `commit` API by requiring a function which will handle/save the changes into some external storage in whatever format it needs to, so `TrieUpdate` won't need to know anything about `CryptoHash` or anything chain-/RPC- specific.

**Comment by frol** [comment]
Resolved with #1910
